# Data Understanding

The purpose of this notebook is to examine the migraine dataset, understand its structure, identify variables relevant to migraine severity prediction, and determine any issues that require preprocessing prior to model development.

In [2]:
from pathlib import Path

import pandas as pd
import numpy as np
import requests

# Set project root
PROJECT_ROOT = Path.cwd().parent if Path.cwd().name == "Notebooks" else Path.cwd()

DATA_DIR = PROJECT_ROOT / "Data"

### Load Data

In [3]:
# Path to migraine dataset
MIGRAINE_FILE = DATA_DIR / "Raw" / "korean_migraine_study_translated.xlsx"

# Load patient data
patients = pd.read_excel(
    MIGRAINE_FILE,
    sheet_name="patients62"
)

# Load headache diary
headache = pd.read_excel(
    MIGRAINE_FILE,
    sheet_name="headache diary 1099"
)

# Load total diary
diary = pd.read_excel(
    MIGRAINE_FILE,
    sheet_name="total diary 4579"
)

print(f"Project root: {PROJECT_ROOT}")
print(f"Data directory: {DATA_DIR}")
print(f"Dataset found: {MIGRAINE_FILE.exists()}")

Project root: c:\Users\rebec\Documents\Thesis
Data directory: c:\Users\rebec\Documents\Thesis\Data
Dataset found: True


### Basic Dataset Info for Each Dataframe

In [4]:
patients.shape

patients.head()

patients.info()

patients.describe()

<class 'pandas.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Columns: 185 entries, Unique ID to Total Moderate Exercise Time.1
dtypes: float64(84), object(63), str(38)
memory usage: 190.9+ KB


,Unique ID,Age,Height,Weight,BMI,Migraine Diagnosis (Other),Headache Frequency (days/month),Disease Duration (years),Duration (hours/episode),Headache Intensity (VAS),...,Total Vigorous Exercise Days,Total Moderate Exercise Days,Total Vigorous Exercise Time,Total Moderate Exercise Time,Unnamed: 179,Total Exercise Days.1,Total Vigorous Exercise Days.1,Total Moderate Exercise Days.1,Total Vigorous Exercise Time.1,Total Moderate Exercise Time.1
count,62.000000,64.000000,54.000000,56.000000,52.000000,0.0,64.000000,64.000000,65.000000,64.000000,...,62.000000,62.000000,62.000000,62.000000,0.0,62.000000,62.000000,62.000000,62.000000,62.000000
mean,57.016129,37.255481,159.150051,58.275238,22.430769,NaN,6.413977,9.717502,30.529906,7.561376,...,2.258065,14.145161,128.064516,849.677419,NaN,3.048387,0.387097,2.758065,23.870968,198.548387
std,40.388885,9.246667,22.199575,12.008852,3.016981,NaN,4.981086,8.049724,25.998631,1.480766,...,7.561628,17.307107,399.265057,1277.022623,NaN,4.403528,1.029958,4.295196,60.985213,505.331726
min,2.000000,8.641090,7.083509,10.265177,16.400000,NaN,1.000000,1.000000,0.032258,1.266770,...,0.000000,0.000000,0.000000,0.000000,NaN,0.000000,0.000000,0.000000,0.000000,0.000000
25%,26.250000,32.750000,157.000000,52.000000,20.550000,NaN,3.000000,3.750000,12.000000,7.000000,...,0.000000,0.250000,0.000000,15.000000,NaN,0.000000,0.000000,0.000000,0.000000,0.000000
50%,49.000000,37.854839,160.000000,55.500000,21.900000,NaN,4.500000,7.000000,24.000000,8.000000,...,0.000000,10.000000,0.000000,400.000000,NaN,2.000000,0.000000,1.000000,0.000000,50.000000
75%,83.750000,42.250000,165.750000,65.250000,24.250000,NaN,8.000000,14.250000,48.000000,8.000000,...,1.000000,19.000000,27.500000,1005.000000,NaN,4.000000,0.000000,4.000000,0.000000,185.000000
max,155.000000,56.000000,184.000000,86.000000,32.500000,NaN,30.000000,30.000000,96.000000,10.000000,...,49.000000,79.000000,2300.000000,5940.000000,NaN,28.000000,5.000000,28.000000,240.000000,3800.000000


In [5]:
patients.head()

,Unique ID,Registration No.,Study No.,Sex,Age,Hospital,Height,Height Unknown,Weight,Weight Unknown,...,Total Vigorous Exercise Days,Total Moderate Exercise Days,Total Vigorous Exercise Time,Total Moderate Exercise Time,Unnamed: 179,Total Exercise Days.1,Total Vigorous Exercise Days.1,Total Moderate Exercise Days.1,Total Vigorous Exercise Time.1,Total Moderate Exercise Time.1
0,2.0,cmc-0001,M-001,Female,20.0,Uijeongbu,155.0,NaN,56.0,NaN,...,0.0,0.0,0.0,0.0,NaN,0.0,0.0,0.0,0.0,0.0
1,3.0,cmc-0002,M-002,Female,38.0,Uijeongbu,155.0,NaN,53.0,NaN,...,0.0,6.0,0.0,190.0,NaN,1.0,0.0,1.0,0.0,30.0
2,5.0,cmc-0003,M-003,Female,20.0,Uijeongbu,155.0,NaN,55.0,NaN,...,0.0,1.0,0.0,120.0,NaN,1.0,0.0,1.0,0.0,120.0
3,6.0,cmc-0004,M-004,Female,37.0,Uijeongbu,164.0,NaN,52.0,NaN,...,0.0,4.0,0.0,280.0,NaN,1.0,0.0,1.0,0.0,40.0
4,8.0,cmc-0006,CM-001,Female,46.0,Uijeongbu,157.0,NaN,52.0,NaN,...,0.0,0.0,0.0,0.0,NaN,0.0,0.0,0.0,0.0,0.0


In [6]:
headache.shape

headache.head()

headache.info()

headache.describe()

<class 'pandas.DataFrame'>
RangeIndex: 1100 entries, 0 to 1099
Columns: 101 entries, No. to Headache Ongoing (Y/N)
dtypes: float64(58), int64(25), object(1), str(17)
memory usage: 868.1+ KB


,No.,Start Time,Date,Intensity (Mild vs Moderate/Severe),Index (VAS Style Score),Pain Left Side,Pain Ride Side,Pain Center,Pain Both Sides,Pain Around Eyes,...,Other Factor: Total,Relief Factor: Sleep,Relief Factor: Rest,Relief Factor: Massage/Stretching,Relief Factor: Exercise,Relief Factor: Other,Relief Factor: Total,Vigorous Exercise,Moderate Exercise,Total Exercise
count,1099.000000,1.099000e+03,1.099000e+03,1067.00000,993.000000,1099.000000,1099.000000,1099.000000,1099.000000,1099.000000,...,1100.000000,1099.000000,1099.000000,1099.000000,1099.000000,1099.000000,1099.000000,1099.000000,1099.000000,1099.000000
mean,1452.321201,2.014298e+11,2.014298e+07,0.48641,4.409869,0.341219,0.226570,0.219290,0.322111,0.156506,...,0.603636,0.316652,0.482257,0.045496,0.035487,0.151956,1.031847,0.847134,0.969973,1.817106
std,1040.021298,3.724547e+07,3.724548e+03,0.50005,2.201581,0.474335,0.418802,0.413954,0.467498,0.363500,...,10.016328,0.465382,0.499913,0.208484,0.185091,0.359142,0.817547,0.360022,0.170740,0.420597
min,8.000000,2.014081e+11,2.014081e+07,0.00000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,617.500000,2.014101e+11,2.014101e+07,0.00000,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,2.000000
50%,1198.000000,2.014111e+11,2.014111e+07,0.00000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,1.000000,2.000000
75%,2091.500000,2.014122e+11,2.014122e+07,1.00000,6.000000,1.000000,0.000000,0.000000,1.000000,0.000000,...,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,2.000000,1.000000,1.000000,2.000000
max,5103.000000,2.015041e+11,2.015041e+07,1.00000,10.000000,1.000000,1.000000,1.000000,1.000000,1.000000,...,332.000000,1.000000,1.000000,1.000000,1.000000,1.000000,5.000000,1.000000,1.000000,2.000000


In [7]:
diary.shape

diary.head()

diary.info()

diary.describe()

<class 'pandas.DataFrame'>
RangeIndex: 4579 entries, 0 to 4578
Columns: 103 entries, No. to Headache Ongoing (Y/N)
dtypes: float64(73), int64(15), object(1), str(14)
memory usage: 3.6+ MB


,No.,Preventive Medication Use,Start Time,Start Time.1,End Time.1,Duration (min),Date,Intensity (Mild vs Moderate/Severe),Index (VAS Style Score),Pain Left Side,...,Relief Factor: Rest,Relief Factor: Massage/Stretching,Relief Factor: Exercise,Relief Factor: Other,Relief Factor: Total,Vigorous Exercise,Moderate Exercise,Total Exercise,Vigorous Exercise (min),Moderate Exercise (min)
count,4579.000000,4579.000000,1.098000e+03,1098.000000,1032.000000,1099.000000,4.579000e+03,1066.000000,993.000000,1098.000000,...,1098.000000,1098.000000,1098.000000,1098.000000,4579.000000,4578.000000,4578.000000,4579.000000,1025.000000,1025.000000
mean,1640.694912,0.404674,2.014298e+11,11.091075,18.732558,474.555960,2.014344e+07,0.486867,4.409869,0.340619,...,0.482696,0.045537,0.035519,0.152095,0.247434,0.809305,0.968545,1.777462,7.746341,51.551220
std,1110.307741,0.490882,3.725813e+07,5.276744,4.931754,345.013124,3.999889e+03,0.500062,2.201581,0.474133,...,0.499928,0.208574,0.185172,0.359276,0.595354,0.392892,0.174562,0.437494,22.523730,43.531077
min,3.000000,0.000000,2.014081e+11,0.000000,1.000000,0.000000,2.014081e+07,0.000000,1.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,744.500000,0.000000,2.014101e+11,8.000000,15.000000,180.000000,2.014102e+07,0.000000,3.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,2.000000,0.000000,20.000000
50%,1470.000000,0.000000,2.014111e+11,11.000000,20.000000,366.000000,2.014112e+07,0.000000,4.000000,0.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,2.000000,0.000000,50.000000
75%,2364.000000,1.000000,2.014122e+11,15.000000,23.000000,689.000000,2.015010e+07,1.000000,6.000000,1.000000,...,1.000000,0.000000,0.000000,0.000000,0.000000,1.000000,1.000000,2.000000,0.000000,60.000000
max,5140.000000,1.000000,2.015041e+11,23.000000,23.000000,1439.000000,2.015041e+07,1.000000,10.000000,1.000000,...,1.000000,1.000000,1.000000,1.000000,5.000000,1.000000,1.000000,2.000000,240.000000,300.000000


### Variable Descriptions

In [8]:
import numpy as np

# Variables grouped according to their role in the study

target_variables = {
    "Severity",
    "Intensity (Mild vs Moderate/Severe)",
    "Index (VAS Style Score)"
}

identifier_variables = {
    "No.",
    "Registration No.",
    "Unique ID"
}

time_variables = {
    "Date",
    "Start Time",
    "Start Time.1",
    "End Time",
    "End Time.1",
    "Duration",
    "Duration (min)"
}

physiological_variables = {
    "Preventive Medication Use",
    "Acute Medication Use",
    "Internal Factor: Stress",
    "Internal Factor: Oversleeping",
    "Internal Factor: Sleep Deprivation",
    "Internal Factor: Exercise",
    "Internal Factor: Lack of Exercise",
    "Internal Factor: Physical Fatigue",
    "Internal Factor: Menstrual Cycle: Menstruation",
    "Internal Factor: Menstrual Cycle: Ovulation",
    "Internal Factor: Excessive Emotional Change",
    "Other Factor: Irregular Meals (fasting, etc.)",
    "Exercise (Y/N)",
    "Vigorous Exercise",
    "Moderate Exercise",
    "Total Exercise",
    "Vigorous Exercise (min)",
    "Moderate Exercise (min)"
}

environmental_variables = {
    "External Factor: Weather/Temperature Change",
    "External Factor: Excessive Sunlight",
    "External Factor: Noise",
    "External Factor: Inappropriate Lighting",
    "External Factor: Specific Smell (cosmetics, perfume, etc.)"
}

outcome_or_post_migraine_variables = {
    "Headache-free Day",
    "Headache Ongoing (Y/N)",
    "Impacts Daily Life",
    "Impacts Daily Life.1",
    "Missed Work/School",
    "Productivity Reduced by Half or More",
    "Total Impact on Work or School",
    "Unable to Do Housework at All",
    "Housework Efficiency Reduced by Half or More",
    "Total Impact at Home",
    "Social/Leisuire: Planned but Unable to Participate",
    "No Effect from Acute Medication",
    "Headache Reduced by Half or More with Acute Medication",
    "Headache Resolved with Acute Medication"
}

def assign_variable_category(column_name):
    if column_name in target_variables:
        return "Target Variable"
    if column_name in identifier_variables:
        return "Identifier Variable"
    if column_name in time_variables:
        return "Date/Time"
    if column_name in physiological_variables:
        return "Physiological"
    if column_name in environmental_variables:
        return "Environmental"
    if column_name in outcome_or_post_migraine_variables:
        return "Outcome/Post-Migraine"
    if column_name.startswith("Pain"):
        return "Headache characteristic"
    if "Headache Type" in column_name or "Headache" in column_name:
        return "Headache characteristic"
    if "Accompanying Symptom" in column_name:
        return "Associated symptom"
    if column_name.startswith("Relief Factor"):
        return "Relief factor"
    if column_name.startswith("Internal Factor"):
        return "Physiological"
    if column_name.startswith("External Factor"):
        return "Environmental"
    if column_name.startswith("Other Factor"):
        return "Other trigger"
    if "Medication" in column_name or column_name in {
        "Tylenol (Acetaminophen)",
        "Combination Drug",
        "Triptan",
        "Other Medication Name"
    }:
        return "Medication"
    return "Other Variable"

variable_descriptions = {
    "No.": "Sequential record number.",
    "Registration No.": "Participant registration number.",
    "Unique ID": "Unique participant identifier.",
    "Date": "Date associated with the diary record.",
    "Start Time": "Recorded starting time of the headache episode.",
    "End Time": "Recorded ending time of the headache episode.",
    "Duration": "Recorded duration of the headache episode.",
    "Duration (min)": "Duration of the headache episode measured in minutes.",

    "Severity": (
        "Reported migraine or headache severity. This is the primary "
        "outcome variable for severity prediction."
    ),
    "Intensity (Mild vs Moderate/Severe)": (
        "Binary grouping that distinguishes mild headaches from "
        "moderate or severe headaches."
    ),
    "Index (VAS Style Score)": (
        "Numeric pain-intensity measure based on a visual analog-style scale."
    ),

    "Preventive Medication Use": (
        "Indicates whether preventive migraine medication was used."
    ),
    "Acute Medication Use": (
        "Indicates whether acute medication was used in response to a headache."
    ),
    "Internal Factor: Stress": (
        "Indicates whether stress was reported as a trigger associated "
        "with the headache episode."
    ),
    "Internal Factor: Oversleeping": (
        "Indicates whether oversleeping was reported as a headache trigger."
    ),
    "Internal Factor: Sleep Deprivation": (
        "Indicates whether insufficient sleep was reported as a headache trigger."
    ),
    "Internal Factor: Exercise": (
        "Indicates whether exercise was reported as a headache trigger."
    ),
    "Internal Factor: Lack of Exercise": (
        "Indicates whether lack of exercise was reported as a headache trigger."
    ),
    "Internal Factor: Physical Fatigue": (
        "Indicates whether physical fatigue was reported as a headache trigger."
    ),
    "Internal Factor: Menstrual Cycle: Menstruation": (
        "Indicates whether menstruation was reported as a headache trigger."
    ),
    "Internal Factor: Menstrual Cycle: Ovulation": (
        "Indicates whether ovulation was reported as a headache trigger."
    ),
    "Internal Factor: Excessive Emotional Change": (
        "Indicates whether a substantial emotional change was reported "
        "as a headache trigger."
    ),
    "Other Factor: Irregular Meals (fasting, etc.)": (
        "Indicates whether fasting, meal skipping, or irregular meal timing "
        "was reported as a headache trigger."
    ),

    "External Factor: Weather/Temperature Change": (
        "Indicates whether a change in weather or temperature was reported "
        "as a headache trigger."
    ),
    "External Factor: Excessive Sunlight": (
        "Indicates whether excessive sunlight was reported as a headache trigger."
    ),
    "External Factor: Noise": (
        "Indicates whether environmental noise was reported as a headache trigger."
    ),
    "External Factor: Inappropriate Lighting": (
        "Indicates whether lighting conditions were reported as a headache trigger."
    ),
    "External Factor: Specific Smell (cosmetics, perfume, etc.)": (
        "Indicates whether exposure to a particular smell was reported "
        "as a headache trigger."
    ),

    "Exercise (Y/N)": (
        "Indicates whether the participant exercised during the relevant period."
    ),
    "Vigorous Exercise": "Recorded amount or occurrence of vigorous exercise.",
    "Moderate Exercise": "Recorded amount or occurrence of moderate exercise.",
    "Total Exercise": "Combined measure of exercise activity.",
    "Vigorous Exercise (min)": (
        "Amount of vigorous exercise measured in minutes."
    ),
    "Moderate Exercise (min)": (
        "Amount of moderate exercise measured in minutes."
    ),

    "Headache-free Day": (
        "Indicates whether the diary date was classified as headache-free."
    ),
    "Headache Ongoing (Y/N)": (
        "Indicates whether a headache was ongoing at the time of the record."
    )
}

data_dictionary = pd.DataFrame({
    "Variable": diary.columns,
    "Current Type": diary.dtypes.astype(str).values,
    "Category": [
        assign_variable_category(column)
        for column in diary.columns
    ],
    "Description": [
        variable_descriptions.get(
            column,
            "Description to be reviewed and added."
        )
        for column in diary.columns
    ],
    "Non-Missing Values": diary.notna().sum().values,
    "Missing Values": diary.isna().sum().values,
    "Percent Missing": (
        diary.isna().mean().mul(100).round(2).values
    ),
    "Unique Values": diary.nunique(dropna=True).values
})

data_dictionary.head(20)

,Variable,Current Type,Category,Description,Non-Missing Values,Missing Values,Percent Missing,Unique Values
0,No.,int64,Identifier Variable,Sequential record number.,4579,0,0.00,3157
1,Registration No.,str,Identifier Variable,Participant registration number.,4579,0,0.00,63
2,Unique ID,str,Identifier Variable,Unique participant identifier.,4579,0,0.00,64
3,Preventive Medication Use,int64,Physiological,Indicates whether preventive migraine medicati...,4579,0,0.00,2
4,Start Time,float64,Date/Time,Recorded starting time of the headache episode.,1098,3481,76.02,1065
5,Start Time.1,float64,Date/Time,Description to be reviewed and added.,1098,3481,76.02,24
6,End Time,object,Date/Time,Recorded ending time of the headache episode.,1038,3541,77.33,814
7,End Time.1,float64,Date/Time,Description to be reviewed and added.,1032,3547,77.46,21
8,Duration,str,Date/Time,Recorded duration of the headache episode.,1099,3480,76.00,378
9,Duration (min),float64,Date/Time,Duration of the headache episode measured in m...,1099,3480,76.00,377


In [9]:
research_variables = data_dictionary[
    data_dictionary["Category"].isin(
        ["Target", "Physiological", "Environmental", "Medication", "Date/Time"]
    )
].reset_index(drop=True)

research_variables

,Variable,Current Type,Category,Description,Non-Missing Values,Missing Values,Percent Missing,Unique Values
0,Preventive Medication Use,int64,Physiological,Indicates whether preventive migraine medicati...,4579,0,0.00,2
1,Start Time,float64,Date/Time,Recorded starting time of the headache episode.,1098,3481,76.02,1065
2,Start Time.1,float64,Date/Time,Description to be reviewed and added.,1098,3481,76.02,24
3,End Time,object,Date/Time,Recorded ending time of the headache episode.,1038,3541,77.33,814
4,End Time.1,float64,Date/Time,Description to be reviewed and added.,1032,3547,77.46,21
5,Duration,str,Date/Time,Recorded duration of the headache episode.,1099,3480,76.00,378
6,Duration (min),float64,Date/Time,Duration of the headache episode measured in m...,1099,3480,76.00,377
7,Date,int64,Date/Time,Date associated with the diary record.,4579,0,0.00,245
8,Acute Medication Use,str,Physiological,Indicates whether acute medication was used in...,1085,3494,76.30,2
9,Tylenol (Acetaminophen),float64,Medication,Description to be reviewed and added.,169,4410,96.31,5


In [10]:
data_dictionary.to_excel(
    "migraine_data_dictionary.xlsx",
    index=False
)

In [11]:
diary["Severity"].value_counts()

Severity
Mild        547
Moderate    402
Severe      117
Name: count, dtype: int64

### Initial Observations

- The diary dataset consists of approximately 4500 daily observations (4579 exact) from 62 migraine patients. Close to 1100 headache/migraine events (1099 exact) were recorded.
- The target variable consists of three severity categories: mild, moderate, and severe. Multiple physiological and environmental predictors were used, such as stress, sleep, menstrual cycle, exercise, dietary habits, medication use, excessive light, excessive smells, and loud sound. Self reported weather as a trigger was also used as a predictor, but actual weather conditions were not included.
- Preliminary examination reveals:
    - There are missing values related to the presence/absence of a headache/migraine (~1099 present/~3480 absent)
    - Any missing values that do not fall into this pattern should the be the highest concern when handling missing values
    - 2 start times (Start Time and Start Time.1) and 2 end times (End Time and End Time.1). Time.1 is from diary dataset and is the hour-time the migraine began and ended. The other Start/End Time is date/time format
    - Some variables in Date/Time category are not the same type (float, object, int). Address this during preprocessing
    - Target variables are unbalanced with severe having a much smaller representation than mild and moderate:
        - Mild: 547
        - Moderate: 402
        - Severe: 117
        - This distribution is to be expected. **Not sure how to address this throughout the project**